sdk/src/types/block/mana/parameters.rs

```diff
@@ -196,6 +196,43 @@ impl ProtocolParameters {
                 - (c >> mana_parameters.decay_factors_exponent())
         })
     }
+
+    pub fn slots_until_generated(
+        &self,
+        generation_amount: u64,
+        stored_mana: u64,
+        required_mana: u64,
+    ) -> Result<u32, ManaError> {
+        if required_mana == 0 {
+            return Ok(0);
+        }
+        let mut num_slots = 0;
+        let mana_generated_per_epoch = self
+            .mana_parameters()
+            .generate_mana(generation_amount, self.slots_per_epoch());
+        if mana_generated_per_epoch == 0 {
+            return Err(ManaError::InsufficientGenerationAmount);
+        }
+        let mut required_mana_remaining = required_mana;
+        loop {
+            // Get the minimum number of slots required to achieve the needed mana (i.e. not including decay)
+            num_slots +=
+                u32::try_from(1 + (required_mana_remaining * self.slots_per_epoch() as u64) / mana_generated_per_epoch)
+                    .map_err(|_| ManaError::InsufficientGenerationAmount)?;
+            // Get the actual values after than many slots
+            let decayed_mana = stored_mana - self.mana_with_decay(stored_mana, 0, num_slots)?;
+            let generated_mana = self.generate_mana_with_decay(generation_amount, 0, num_slots)?;
```

Comment:
Just noticed there is ManaParameter::annual_decay_factor_percentage shouldn't this have an effect on the decay depending on the slot range? Currently it doesn't seem to make any difference, but do we need an issue to handle this?

Reply:
I don't think so? But maybe I don't understand it

Reply:
Okay, that's fixable